### ISSUE TITLE
UX for Information Markers

### ISSUE THREAD

**Issue by IanMayo** [serge-web/serge#1522]
The `Game Controller` can now add and edit info markers.  Background [here](https://docs.google.com/presentation/d/1AOGfEQuSYsRyvAYUe4_x3E_bvu15KJ8KpzqDyDAMh-Q/edit#slide=id.g113e814bd03_0_0)

The form will probably be similar to the Perception Form:
![image](https://user-images.githubusercontent.com/1108513/164452342-a981c38f-b7d0-495d-9381-5626273903b9.png)

Here is the flow:
- If logged in as Game Controller, an `Add information marker` icon is shown in the map-control-bar (see below)
- When this button is clicked, a new marker is placed at the centre of the current view
- Once added, the marker is selected, and the `Edit marker form` is shown.
- This form allows the Game Controller to change these properties:
    - label
    - description (text-area)
    - icon (from drop-down list. These aren't platform-types, they are a separate list that are maintained in a new Game Admin tab)
    - visible to (list of check-boxes, one per force)
    - radius (integer, default value zero)
    - Color (this may be the color of a force, or any other color - since the Game Control may specify that icons of particular color relate to same subjects, such as `marine mammals`, `civilian interactions`, `weather features`)
- Once edits are complete, the Game Controller clicks on `Save`

![image](https://user-images.githubusercontent.com/1108513/164451485-987d3b9f-5229-4b02-927c-2ef483a0dcda.png)

**Comment by ShendMikullovci** [comment]
>   * icon (from drop-down list. These aren't platform-types, they are a separate list that are maintained in a new Game Admin tab)

What icons would these be? Similar to the ones we've discussed in #1339 ?

**Comment by ShendMikullovci** [comment]
> The `Game Controller` can now add and edit info markers. Background [here](https://docs.google.com/presentation/d/1AOGfEQuSYsRyvAYUe4_x3E_bvu15KJ8KpzqDyDAMh-Q/edit#slide=id.g113e814bd03_0_0)

Was just reading the slides and I noticed that you had initially requested a draggable marker displayed on the map. While it's a necessary feature, could we include an "undo movement" feature for that? I can already see cases of accidentally dragging the pin to a different location and needing it to go back. A simple CTRL + Z maybe? That way we could also show it in the "planned steps" or "history".

**Comment by IanMayo** [comment]
> > * icon (from drop-down list. These aren't platform-types, they are a separate list that are maintained in a new Game Admin tab)
> 
> What icons would these be? Similar to the ones we've discussed in #1339 ?

Duh - yes, those icons.  Sorry @ShendMikullovci - I'd forgotten that you had already done the design for this form.

![image](https://user-images.githubusercontent.com/1108513/164649342-7f45ce75-3e44-4d45-9e26-c4fcfe789daf.png)

Aaah, not quite.  We need fields for color and radius, as described above.

The presentation includes some examples of the icons that could be used for Information Markers.  I'll get our artist to create some, to get the clients started, and to help in reviewing the feature.

**Comment by IanMayo** [comment]
@ShendMikullovci - I've realised we also need UI for a Game Admin page where the Game Controller can maintain the information marker icons.  They'll need to be able to upload/change the SVG file for an icon (like they do for a `force`), and specify the name/title of the icon.

**Comment by IanMayo** [comment]
>  I can already see cases of accidentally dragging the pin to a different location and needing it to go back

The way it has been implemented, this is actually less likely. First the Game Control has to click on the marker to select it, then it becomes draggable.  So, for now I think we're ok without `ctrl-z` undo.

**Comment by ShendMikullovci** [comment]
Here's an idea for the in-game design: 

![Info Marker UX](https://user-images.githubusercontent.com/85491794/165157643-5f24c32a-2f77-4825-8fd4-e37e712dbd25.png)

We maintain uniformity by keeping the structure of the perception form. Label on the top left, icon on the top right.

The description follows the label and icon, in its own white cloud (white = editable, as usual). 

I took the liberty of changing our visibility display where we provide the forces in clickable circles. The circles that have been clicked are outlined in white, making the info marker thus visible to those forces.

Added color and radius to their own little grey cloud. Both editable for GC. 
Side question: Would radius and color as variables be shown for players outside of GC/Umpire? I'm not sure if there's value to that.

Cancel and Save buttons in their adequate colors. Save is emboldened for emphasis, Cancel is italicized and slim.

**Comment by ShendMikullovci** [comment]
As for the admin page, here's a design idea: 

![Info Marker Admin Page](https://user-images.githubusercontent.com/85491794/165159905-e394aaaf-794c-4fd1-a1f0-0f5d56db2a36.png)

Wanted to maintain uniformity here as well with the rest of the admin pages. 

Couple of things I changed are:
1. "Change Icon" & "Change Color" should behave like "Add Speed" or "Add Condition" buttons: Light grey on standby, Bold grey on hover.
2. Using the visibility design from the in-game design, instead of using the usual game channel row. This introduces a difference between the two pages and I'm OK with not going forward with it, but I think it's more important to link force color to info markers as much as we can, due to their emphasis on the mapping page.

Let me know your thoughts.

**Comment by IanMayo** [comment]
Thanks for those @ShendMikullovci .

Popup dialog first:
- when we have lots of forces, we can't just rely on colors. But, we've had to solve that elsewhere already. We already have `visible-to` editor. That gives a compact list of names & colors. We will modify that control to present a drop-down menu if there are more than, say 6 forces.  
- color is presented to the players as the visible color of the icon. Radius will be displayed as a semi-transparent shaded area beneath the icon. They could count the rings in the radius if they wanted to.
- I think that _by-default_ we should allow two lines for the `description` box, just to demonstrate to the Game Designer that it can be a multi-line entry.
- Aah, one tricky issue. The color could either be the color of a force (where the event relates to one force), or any custom color (if the Game Designer comes up with colors for different types of concept). So, I think we need some kind of   fancy solution.   _Maybe_ a drop-down list of forces, plus `custom`.  When `custom` is selected, the adjacent free-form color selector is used.

Now the admin page:
- I _Don't-think_ we need an admin page for `information-markers`, all maintenance can be done from the mapping page.
- But, we do need an admin page for `annotation-icons` (or maybe `marker-icons`).  This is where the game designer adds icons for new annotations.  It just needs name and icon.

**Comment by ShendMikullovci** [comment]
>Radius will be displayed as a semi-transparent shaded area beneath the icon. They could count the rings in the radius if they wanted to.

Can you clarify this? What rings? How large is the area? 

>Aah, one tricky issue. The color could either be the color of a force (where the event relates to one force)

That's what I thought the color would represent initially. The "custom" idea sounds like a valid solution, as long as color really is a necessary component for the markers. How crucial is it? 

>I Don't-think we need an admin page for information-markers, all maintenance can be done from the mapping page.
But, we do need an admin page for annotation-icons (or maybe marker-icons). This is where the game designer adds icons for new annotations. It just needs name and icon.

Sorry, I thought you meant a fresh admin page when you mentioned this: 

>icon (from drop-down list. These aren't platform-types, they are a separate list that are maintained in a new Game Admin tab)

**Comment by ShendMikullovci** [comment]
Here's an idea for the pop-up:

![Info Markers Pop-up](https://user-images.githubusercontent.com/85491794/165976791-0bd4166f-770b-4225-ad90-a1bfb116b67f.png)

Force names in their appropriate color presented in check-boxes. Color is removed from the pop-up since it only pertains to the icon. Essentially, the icon shown opposite the label could be in different colors I suppose.
Not really sure I understand the radius yet, so I stuck to the value design. Let me know how we could modify this.

**Comment by ShendMikullovci** [comment]
As for the Admin Page for the Markers, I thought we should stick to the design flow we present for other pages:

![Admin Page Markers](https://user-images.githubusercontent.com/85491794/165977548-9e2c9b37-7169-496f-a189-d3437b5b88e1.png)

Here, I chose to center the elements a little since they're on a virtually empty page.

I also had the idea of marker categories: I realized that the user might need several markers of the same category (for example, different markers to for weather annotations). We could present them with the possibility of creating marker categories, presented in a similar way to assets in the forces page: 

![Marker Categories 1](https://user-images.githubusercontent.com/85491794/165979043-2ba35706-76ff-4369-b26b-46174d1752e1.png)

Similar to the list of assets, we show a list of markers relevant to a category. Hitting Add New Marker would open the following window: 

![Marker Categories 2](https://user-images.githubusercontent.com/85491794/165979442-ac3c357a-db56-4442-ae29-5a7208597d78.png)

The user could then add the icon, the description and the label to the marker. 

Let me know your thoughts.

**Comment by IanMayo** [comment]
## Marker editor
* I find the blue, red and pink text labels lack contrast. I'd be happier to stick with "normal" colored labels for the forces, but to have the checkbox in the relevant shade.  The square checkbox outline would still be legible.
*  I'm pretty sure the force labels should be left-aligned
* The `radius` refers to the number of additional cells this event relates to. It's an integer. A zero value means there is not an additional shaded outline behind the event marker.   If the event is "ship on fire (Smoke)", it may be that the marker has a central location, but the semi-transparent block behind it extends out for a further 2 cells in each direction - so it has a `radius` of `2`.

## Marker Icons panel
* Sure, simple centered icon is fine.  Maybe we should use luxury of whitespace to include more guidance on the page - such as drag/drop in a new image file to update the icon
* Right now we don't need categories. I think we just have 7 icons for the markers
* We don't have the `description` on this page.  That is not specified when the Umpire has put the marker into the game. So, these are `icons`. The entities that get added to the game are `Map Annotations` (which can include a marker).

**Comment by ShendMikullovci** [comment]
> * I'd be happier to stick with "normal" colored labels for the forces, but to have the checkbox in the relevant shade.  The square checkbox outline would still be legible.

Great idea. Wasn't sure if that could be done to be honest... 
Here's a look: 

![Info Markers Pop-up New](https://user-images.githubusercontent.com/85491794/167420885-0831177b-1e52-488d-ac7b-bbdec6c124fc.png)
 
> ## Marker Icons panel
> * Sure, simple centered icon is fine.  Maybe we should use luxury of whitespace to include more guidance on the page - such as drag/drop in a new image file to update the icon

Given that the "Change Icon" feature is already present in this format for forces, I suggest we stick to this way of presenting it. We could have the cursor reveal a line of text or two when hovering over the icon and feature like so: 

![Admin Page Markers New](https://user-images.githubusercontent.com/85491794/167421127-f383cbae-3887-4f5c-b7b7-9c2540d1a511.png)

Let me know what you think.

**Comment by IanMayo** [comment]
Hello @ShendMikullovci - I have a new question.  In `Adjudication phase` - all marker changes get `grouped up`, and processed on `Submit new State of World`.  But, in `Planning Phase`, changes to info markers are sent out immediately. This is useful in case the umpire has forgotten to make a marker visible to one force, or if they need to change the label/description.

This issue is about that second version.  Currently, if umpire selects a Marker, and drags it, then the message gets sent out instantly to all players.  If the icon is changed, this is another update.

This may be a mistake. The umpire may wish to change the location _And_ the icon.  This would allow this UX:
- click marker to select
- edit panel displayed
- umpire can drag marker
- umpire either drags marker, or edits panel, (or both)
- umpire clicks on `Save` - and new marker location/icon/label sent out immediately

The above strategy means that only one update get sent out.

**Comment by ShendMikullovci** [comment]
I was thinking about this and realized that the label and icon don't look editable at all so I decided to go at it again and play with shadows:

![Info Markers Shadowed Label and Icon](https://user-images.githubusercontent.com/85491794/167946850-a04e503b-f0cf-476a-b966-026f9965cd69.png)

The shadow sort of brings them out of the display, and is a recurring theme on most clickable objects on screens so it should be intuitive for the users. I also realized that since we can't really use colors as easily as we thought to distinguish the markers from other elements, shadows could be our way around that problem.

The only UX issue I have here is that clicking the marker now leads to two functions: 
1. Relocation
2. Editing

These are two fairly different processes that may interfere with one another. Can we have the marker selected for relocation in one click, and have the edit panel displayed only after a double click? That way, if the user only wants to move the marker, they don't have to deal with the edit panel and can simply click once and move the marker around. Double clicking to edit also places more intention in the action.

**Comment by IanMayo** [comment]
Thanks @ShendMikullovci,
I'm not convinced that the shadow suggests to the player that something is editable.

How about we display the label and icon in their editable form?  So, the label "field" is shown with an underline plus a cursor, and the icon is shown with a drop-down menu (to allow switching to another icon).  The other fields on the panel are displayed in their "being edited" format - but currently for the first two the player must click-to-edit. By starting off in the edit form we demonstrate that they are editable.

For the other issue "different processes for editing a marker", I'm not sure that they are different.  Once a marker is clicked, it becomes editable - either by the form or by dragging.  Clicking on "Save" will save any changes - either by dragging or by editing the form.

**Comment by lebaphi** [comment]
Hi @IanMayo and @ShendMikullovci , do we finalize this feature? 😄

**Comment by ShendMikullovci** [comment]
> How about we display the label and icon in their editable form? So, the label "field" is shown with an underline plus a cursor, and the icon is shown with a drop-down menu (to allow switching to another icon). 

Let's give it a go! I'll get on the dropdown design.

> Once a marker is clicked, it becomes editable - either by the form or by dragging. Clicking on "Save" will save any changes - either by dragging or by editing the form.

Yes, but do we intend for the form to pop up every time the marker is clicked? That's where I suggested we have the form only pop up after the marker is _double clicked_, that way the form only occupies visual space when necessary. 
Single clicks = relocation (the marker would highlight but the form wouldn't appear), double clicks = data editing (the form would appear taking up visual space in the map).

**Comment by ShendMikullovci** [comment]
Here's some ideas on how the drop-down can look:

![Dropdowns](https://user-images.githubusercontent.com/85491794/168561864-98b81a44-e3d8-40b5-ba7a-2d8e0b8cca47.png)

Granted, some of the icons might be self-explanatory and need no label, I thought it could still be valuable. After all, we should reduce as much quantity of information from the user's working memory as possible.

Here's how they would look in the form:

![Info Marker Icon Drop Down](https://user-images.githubusercontent.com/85491794/168562105-20fcb59f-6a45-4301-a657-2a147cce1267.png)

![Info Marker Icon Drop Down 2](https://user-images.githubusercontent.com/85491794/168562135-1389ab3c-c5e0-4fc1-981b-cec90fa71bbb.png)

I also added the downward arrow on top of the underscore like we do for other drop-downs to maintain uniformity and reinforce editability. 

Let me know what you think.

**Comment by IanMayo** [comment]
Thanks @ShendMikullovci :-)

Yes, names alongside the icons would be useful. I expect @lebaphi will handle this, but for the Label, the underline should span the full width of the control (rather than the width of the current text).

**Comment by ShendMikullovci** [comment]
> but for the Label, the underline should span the full width of the control (rather than the width of the current text).

Nicely spotted. 

![Info Marker Drop Down Final](https://user-images.githubusercontent.com/85491794/168566419-4d4d6e84-19e4-4d9b-b8c9-66b54243c3fd.png)

**Comment by IanMayo** [comment]
I've just realised that we can't edit the color of the marker in this form.

I propose we do this in two phases.  

1. Add a new "color" panel, which includes the popup color editor we use in the "Force" admin page.
2. @ShendMikullovci designs a new control that (somehow 😉 ) allows the Game Controller to choose either a force color or one of a palette of colors. Probably a drop-down combo box.

If we get "1" done first, then we can get the feature tested, including on the backend.

For phase 2, I've moved away from a free color choice.  My guess is that the Game Controller may wish to use a set of colors to represent different things, `red/amber/green` for level of threat, for example.  If can be difficult to get the same shade of amber for multiple icons via a color editor.  The fixed palette means the Game Controller can design a set of meanings to different colors.

**Comment by ShendMikullovci** [comment]
> 1. Add a new "color" panel, which includes the popup color editor we use in the "Force" admin page.
> 2. @ShendMikullovci designs a new control that (somehow wink ) allows the Game Controller to choose either a force color or one of a palette of colors. Probably a drop-down combo box.

How about a fill icon? 

![Fill Icon Black White Background](https://user-images.githubusercontent.com/85491794/169043783-933f9f5e-ed5f-4d7a-a498-db2192029ad8.png)

I thought of putting it on the header of the form next to the icon feature, since that's how we present it in the admin page as well. Here's how it looks:

![Info Marker Fill Feature](https://user-images.githubusercontent.com/85491794/169044737-870732a0-b9b4-488a-b46b-5054a01c0ee4.png)

Clicking on it or the color rectangle pops up the new, more modest color editor: 

![Info Marker Fill Feature 2a](https://user-images.githubusercontent.com/85491794/169045319-5d09ac33-2642-4379-bb2c-07cb891556ff.png)

Here, we only show the Force colors currently in use, and a preset of general colors. 

What do you think?

**Comment by IanMayo** [comment]
Sure, nice design @ShendMikullovci .

I'm _tempted_ to try to include the force-name in the colors.  Sometimes there will be multiple blue forces, each with a different shade of blue.

**Comment by ShendMikullovci** [comment]
> Sure, nice design @ShendMikullovci .
> 
> I'm _tempted_ to try to include the force-name in the colors. Sometimes there will be multiple blue forces, each with a different shade of blue.

Thought of it too but assumed you wouldn't want to take up space. I'll add them right away.

**Comment by ShendMikullovci** [comment]
There:

![Info Marker Fill Feature 2a](https://user-images.githubusercontent.com/85491794/169048878-8f82df21-4a16-4e91-800b-8471719b9b4f.png)

P.S. I also suggest we add the fill icon underneath the admin pages color editor rectangle. Currently it just looks like a blank rectangle and gives the user no indication on its service.

**Comment by IanMayo** [comment]
I'm re-opening this. We're most of the way there.

But, could you modify it so that we show the picture as well as the name for the icons please @lebaphi ?

**Comment by IanMayo** [comment]
Hello @lilitkarapetyan  - we do still need the admin page to edit the Marker Icons, please:
https://github.com/serge-web/serge/issues/1522#issuecomment-1121109274

Note: since we have an initial, default set of icons, I suggest we declare these in the `default wargame` document for `create new wargame`.  This initial set of values can be seen here:
https://github.com/serge-web/serge/blob/develop/packages/mocks/watu-wargame-mock.ts#L1263

(oops, specific line number may change once I merge next PR)

**Comment by IanMayo** [comment]
Hello @ShendMikullovci - when you get chance, could you do a review of where we've got to with this?

You should be able to test it from this link: https://serge-review.herokuapp.com/

The functionality is certainly the most important - that the features/capabilities you specified are in there.

**Comment by ShendMikullovci** [comment]
> Hello @ShendMikullovci - when you get chance, could you do a review of where we've got to with this?
> 
> You should be able to test it from this link: https://serge-review.herokuapp.com/
> 
> The functionality is certainly the most important - that the features/capabilities you specified are in there.

It seems to be working well. However, clicking on the marker from a player tab doesn't show the bottom left panel and collapses the top left: 

https://user-images.githubusercontent.com/85491794/177154977-bd450bfe-5f69-46a7-9865-d58987c71414.mp4

**Comment by IanMayo** [comment]
Thanks @ShendMikullovci - the lower-left panel only opens for Game Control - to edit the Info Markers.  For players, they just interact with the marker on the map (plus the tooltip for more details).

For the upper-left panel, we'll have to think about what's consistent. The current collapse behaviour is what we put in for assets & planning/adjudication.  Clicking twice on one collapses the top-left panel to give more room for planning/adjudication.  That's less important for info markers.

**Comment by ShendMikullovci** [comment]
> Thanks @ShendMikullovci - the lower-left panel only opens for Game Control - to edit the Info Markers. For players, they just interact with the marker on the map (plus the tooltip for more details).

In that case, we're good. I think the only thing I'd change is when the user clicks on a marker and the top right collapses, they should only be able to re-open it by directly clicking on it. We don't want the user to confuse the features and functionalities between them.
